VoiceOver/DataModules.lua

```diff
@@ -0,0 +1,160 @@
+        error(format([[Module "%s" contains outdated data format (version %d, expected %d)]], name, moduleVersion, CURRENT_MODULE_VERSION))
+        return
+    end
+
+    module.MODULE_NAME = name
+    module.MODULE_VERSION = moduleVersion
+    module.MODULE_PRIORITY = tonumber(GetAddOnMetadata(name, "X-VoiceOver-DataModule-Priority")) or 0
+    module.MODULE_TITLE = GetAddOnMetadata(name, "Title") or name
+    module.CONTENT_VERSION = GetAddOnMetadata(name, "Version")
+
+    self.registeredModules[name] = module
+    table.insert(self.orderedModules, module)
+
+    -- Order the modules by priority (higher first) then by name (case-sensitive alphabetical)
+    -- Modules with higher priority will be iterated through first, so one can create a module with "overrides" for data in other modules by simply giving it a higher priority
+    table.sort(self.orderedModules, function(a, b)
+        if a.MODULE_PRIORITY ~= b.MODULE_PRIORITY then
+            return a.MODULE_PRIORITY > b.MODULE_PRIORITY
+        end
+        return a.MODULE_NAME < b.MODULE_NAME
+    end)
+end
+
+function DataModules:GetModules()
+    return ipairs(self.orderedModules)
+end
+
+function DataModules:EnumerateAddons()
+    local playerName = UnitName("player")
+    for i = 1, GetNumAddOns() do
+        if GetAddOnMetadata(i, "X-VoiceOver-DataModule-Version") and GetAddOnEnableState(playerName, i) ~= 0 then
+            local name = GetAddOnInfo(i)
+            local mapsString = GetAddOnMetadata(i, "X-VoiceOver-DataModule-Maps")
+            local maps = { }
+            if mapsString then
+                for _, mapString in ipairs({ strsplit(",", mapsString) }) do
+                    local map = tonumber(mapString)
+                    if map then
+                        maps[map] = true
+                    end
+                end
+            end
+            self.availableModules[name] =
+            {
+                Name = name,
+                Maps = maps,
+                LoadOnDemand = IsAddOnLoadOnDemand(name),
+            }
+
+            -- Maybe in the future we can load modules based on the map the player is in (select(8, GetInstanceInfo())), but for now - just load everything
+            if IsAddOnLoadOnDemand(name) then
+                LoadAddOn(name)
+            end
+        end
+    end
+end
+
+function DataModules:GetNPCGossipTextHash(soundData)
+    local npcId = VoiceOverUtils:getIdFromGuid(soundData.unitGuid)
+    local text = soundData.text
+
+    local best_result
+
+    for _, module in self:GetModules() do
+        local data = module.NPCToTextToTemplateHash
+        if data then
+            local npc_gossip_table = data[npcId]
+            if npc_gossip_table then
+                local text_entries = {}
+                for text, _ in pairs(npc_gossip_table) do
+                    table.insert(text_entries, text)
+                end
+
+                local result = VOICEOVER_fuzzySearchBest(text, text_entries)
```

Comment:
fuzzySearchBest already returns the best result, you dont need to rerun this function. It also doesnt mutate text_entries, so we dont need to copy the lookup table for the npc before passing it in.

Reply:
> fuzzySearchBest already returns the best result, you dont need to rerun this function.

It's run once for every module, in case multiple modules provide gossips for the same npc.

> It also doesnt mutate text_entries, so we dont need to copy the lookup table for the npc before passing it in.

I copied your code, you made that `text_entries` :P
I figured it's because npc_gossip_table is an associative array, while your implementation of fuzzy search wants an indexed array. For more performance you could rewrite fuzzy search to iterate through the container's keys instead of values, then it would work with `npc_gossip_table`. However...

I suppose instead of running fuzzy search for every module we can instead collect all gossips for the same npc from all modules into one container and run fuzzy search on all of them at once.

Reply:
```diff
diff --git a/VoiceOver/DataModules.lua b/VoiceOver/DataModules.lua
index 205d3cb..44950d8 100644
--- a/VoiceOver/DataModules.lua
+++ b/VoiceOver/DataModules.lua
@@ -84,28 +84,22 @@ function DataModules:GetNPCGossipTextHash(soundData)
     local npcId = VoiceOverUtils:getIdFromGuid(soundData.unitGuid)
     local text = soundData.text
 
-    local best_result
+    local text_entries = {}
 
     for _, module in self:GetModules() do
         local data = module.NPCToTextToTemplateHash
         if data then
             local npc_gossip_table = data[npcId]
             if npc_gossip_table then
-                local text_entries = {}
-                for text, _ in pairs(npc_gossip_table) do
-                    table.insert(text_entries, text)
-                end
-
-                local result = VOICEOVER_fuzzySearchBest(text, text_entries)
-                if result and (not best_result or result.distance < best_result.distance) then
-                    best_result = result
-                    best_result.hash = npc_gossip_table[result.text]
+                for text, hash in pairs(npc_gossip_table) do
+                    text_entries[text] = text_entries[text] or hash -- Respect module priority, don't overwrite the entry if there is already one
                 end
             end
         end
     end
 
-    return best_result and best_result.hash
+    local best_result = VOICEOVER_fuzzySearchBestKeys(text, text_entries)
+    return best_result and best_result.value
 end
 
 function DataModules:GetQuestLogNPCID(questId)
diff --git a/VoiceOver/FuzzySearch.lua b/VoiceOver/FuzzySearch.lua
index 45bdabb..237165a 100644
--- a/VoiceOver/FuzzySearch.lua
+++ b/VoiceOver/FuzzySearch.lua
@@ -50,6 +50,25 @@ function VOICEOVER_fuzzySearchBest(query, entries)
     return best_result
 end
 
+function VOICEOVER_fuzzySearchBestKeys(query, entries)
+    local best_result = nil
+    local min_distance = math.huge
+
+    for entry, value in pairs(entries) do
+        local distance = levenshtein_distance(query, entry)
+        if distance < min_distance then
+            min_distance = distance
+            best_result = {
+                value = value,
+                text = entry,
+                distance = distance
+            }
+        end
+    end
+
+    return best_result
+end
+
 -- Example usage
 -- local entries = {"apple", "banana", "orange", "pineapple", "grape", "watermelon", "strawberry"}
  
```